cpu/esp32/periph/spi.c

```diff
@@ -0,0 +1,432 @@
+    spi_ll_set_rx_lsbfirst(_spi[bus].periph->hw, false);
+    spi_ll_set_tx_lsbfirst(_spi[bus].periph->hw, false);
+
+    /* acquire and release to set default parameters */
+    spi_acquire(bus, GPIO_UNDEF, SPI_MODE_0, SPI_CLK_100KHZ);
+    spi_release(bus);
+
+    return;
+}
+
+void spi_init_pins(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+
+    /* avoid multiple pin initializations */
+    if (_spi[bus].pins_initialized) {
+        return;
+    }
+    _spi[bus].pins_initialized = true;
+
+    DEBUG("%s bus=%u\n", __func__, bus);
+
+    if (gpio_init(spi_config[bus].sck, GPIO_OUT) ||
+        gpio_init(spi_config[bus].mosi, GPIO_OUT) ||
+        gpio_init(spi_config[bus].miso, GPIO_IN)) {
+        LOG_TAG_ERROR("spi",
+                      "SPI_DEV(%d) pins could not be initialized\n", bus);
+        return;
+    }
+    if (spi_init_cs(bus, spi_config[bus].cs) != SPI_OK) {
+        LOG_TAG_ERROR("spi",
+                      "SPI_DEV(%d) CS signal could not be initialized\n",
+                      bus);
+        return;
+    }
+
+    /* store the usage type in GPIO table */
+    gpio_set_pin_usage(spi_config[bus].sck, _SPI);
+    gpio_set_pin_usage(spi_config[bus].mosi, _SPI);
+    gpio_set_pin_usage(spi_config[bus].miso, _SPI);
+
+    /* TODO  the IO_MUX should be used instead of GPIO matrix routing for
+             lower delays and higher clock rates whenever possible */
+
+    /* connect SCK and MOSI pins to the output signal through the GPIO matrix */
+    esp_rom_gpio_connect_out_signal(spi_config[bus].sck,
+                                    _spi[bus].periph->spiclk_out, false, false);
+    esp_rom_gpio_connect_out_signal(spi_config[bus].mosi,
+                                    _spi[bus].periph->spid_out, false, false);
+
+    /* connect MISO input signal to the MISO pin through the GPIO matrix */
+    esp_rom_gpio_connect_in_signal(spi_config[bus].miso,
+                                   _spi[bus].periph->spiq_in, false);
+}
+
+void spi_deinit_pins(spi_t bus)
+{
+    /* TODO */
+#if IS_USED(MODULE_SPI_RECONFIGURE)
+#error not yet implemented
+#endif
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    DEBUG("%s bus=%u cs=%u\n", __func__, bus, cs);
+
+    assert(bus < SPI_NUMOF);
+
+    /* return if pin is already initialized as SPI CS signal */
+    if (gpio_get_pin_usage(cs) == _SPI) {
+        return SPI_OK;
+    }
+
+    /* check whether CS pin is used otherwise */
+    if (gpio_get_pin_usage(cs) != _GPIO) {
+        return SPI_NOCS;
+    }
+
+    /* initialize the pin */
+    gpio_init(cs, GPIO_OUT);
+    gpio_set(cs);
+
+    /* pin cannot be used for anything else */
+    gpio_set_pin_usage(cs, _SPI);
+
+    return SPI_OK;
+}
+
+static const uint32_t _spi_clk_freq_hz[] = {
+    100 * KHZ(1),   /* SPI_CLK_100KHZ */
+    400 * KHZ(1),   /* SPI_CLK_400KHZ */
+    1 * MHZ(1),     /* SPI_CLK_1KHZ */
+    5 * MHZ(1),     /* SPI_CLK_5MHZ */
+    10 * MHZ(1),    /* SPI_CLK_10MHZ */
```

Comment:
Just drop this and let `clk` be the clock in MHz directly

Reply:
Hm, this change and required subsequent changes cause now compilation errors for ESP32-C3 :worried: